### PR TITLE
Introduce unit tests

### PR DESCRIPTION
**Pull request by marcusvaltonen**
Hi @freddyaboulton!

I am starting a draft merge request here, since this should probably be iterated over. I used pytest and FastAPI's TestClient (httpx in the background), whereas aiortc uses unittest and aiohttp.

For now, I only wrote some dummy tests for ``WebRTCConnectionMixin``. I guess in most cases you really just want to mock the functionality and you do not want/need a complete session with client-server messaging back and forth. Anyhow, here is a rough draft. Let me know your thoughts, and how to set the scope for introducing unittests as well, e.g., what modules should be tested to start with?

Feel free to edit/add to the PR as you wish.

### CODE REVIEW

**Comment by marcusvaltonen** [comment]
This solves #239

**Comment by freddyaboulton** [comment]
Thank you so much @marcusvaltonen ! I will add some more tests soon!

**Comment by marcusvaltonen** [comment]
OK, I will rebase it and add some in the meantime.

**Comment by marcusvaltonen** [comment]
@freddyaboulton I marked some cases with ``pytest.mark.xfail`` that I think requires some extra attention. Perhaps it is a feature, perhaps a bug..

File: test/test_webrtc_connection_mixin.py [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+        assert status == 'failed'
+        assert metadata == {'error': 'concurrency_limit_reached', 'limit': 2}
+
+        await self.close_peer_connection(pc1)
+        await self.close_peer_connection(pc2)
+        await self.close_peer_connection(pc3)
+
+    @pytest.mark.asyncio
+    @pytest.mark.parametrize("test_client_and_stream", [{"concurrency_limit": 1}], indirect=True)
+    @pytest.mark.xfail(reason="No audio or video tracks in peer connections added only datachannels")
```
Comment (marcusvaltonen) [inline test/test_webrtc_connection_mixin.py:249]: This is an interesting case. As long as a datachannel is added to the peer request it is a valid offer; however, they will not be added to ``stream.connections``.

Personally, I believe the best case would be to catch that no valid media stream(s) (depending on modality) is present, and close the connection, i.e., if you have ``modality=audio`` you need to supply an audio stream within a time limit, likewise, if you have ``modality="audio-video"`` both must be present.

File: test/test_webrtc_connection_mixin.py [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+        assert status == 'failed'
+        assert metadata == {'error': 'concurrency_limit_reached', 'limit': 1}
+
+        await self.close_peer_connection(pc1)
+        await self.close_peer_connection(pc2)
+
+    @pytest.mark.asyncio
+    @pytest.mark.parametrize("test_client_and_stream", [{"concurrency_limit": 2}], indirect=True)
+    @pytest.mark.xfail(reason="It is the same webrtc_id so pc1 and pc2 are seen as ONE connection")
+    async def test_concurrency_limit_reached_three_peers_same_id(self, test_client_and_stream):
```
Comment (marcusvaltonen) [inline test/test_webrtc_connection_mixin.py:212]: Here, as long as ``concurrency_limit>1`` and you reuse ``webrtc_id`` you can supply as many peer connections you want since ``self.connections`` has length 1.

Should probably be restricted in some way, alternatively have a limit for how many connections one ``webrtc_id`` can have.

**Comment by marcusvaltonen** [comment]
@freddyaboulton My suggestion is to merge this as is and write Issues for fixing the above found bugs. Other tests for the existing framework can be merged subsequently, as they are ready. New features should include tests. By also adding coverage reports, e.g., CodeCov, one can track the progress.

File: test/test_webrtc_connection_mixin.py (hunk 2 above)
Comment (freddyaboulton) [reply test/test_webrtc_connection_mixin.py]: We should not allow duplicate webrtc_ids. Will push in your branch.

File: test/test_webrtc_connection_mixin.py (hunk 1 above)
Comment (freddyaboulton) [reply test/test_webrtc_connection_mixin.py]: I think we should return the error if a peer connection has been received with a previous id. regardless of whether the tracks have been sent or the modality. Will push as well.

**Comment by freddyaboulton** [comment]
Thank you for this great PR @marcusvaltonen ! I modified the webrtc_mix + tests in light of the corner cases you've found. Also made a change to not test every python version an instead test the min and max version.
